### ISSUE TITLE
`quote_style` warnings

### ISSUE THREAD

**Issue by ibhagwan** [CppCXY/EmmyLuaCodeStyle#94]
Hi again @CppCXY,

Using `quote_style=double` the formatter generates warnings for what it deems "incorrect" code style, for exmaple:
```lua
local foo = 'bar'
```
will generate the warning `' should be "` which is correct in this case.

However, the same warnings are also generated when it's not possible to replace the single quote with double, consider the below:
```lua
local foo = 'b"a"r'
```
will also generate the same warning `' should be "`, but I'm not sure the warning should be generated in this case, for example stylua uses `quote_style = "AutoPreferDouble"` which will replace single quotes with double quotes **only when the inner string does not contain a double quote**.

What do you think?

**Comment by CppCXY** [comment]
formatting will check to see if it can be converted to double quotes, but diagnostics are simpler.
you mentioned `Stylua`, but in the handling of formatting quotes, we are the same.
if you think the diagnostic and formatting goals should be consistent at this point, I'm okay with that

**Comment by ibhagwan** [comment]
> formatting will check to see if it can be converted to double quotes, but diagnostics are simpler.
> you mentioned `Stylua`, but in the handling of formatting quotes, we are the same.
> if you think the diagnostic and formatting goals should be consistent at this point, I'm okay with that

Ty @CppCXY, I’m mentioning stylua just as a reference point, doesn’t mean the behavior should be the same.

My issue with the diagnostics warning here is that it can never be fixed, if you need the double quotes inside the string you will forever have the warning unless you’re willing to use ugly double quote escape `”b\”a\”r”` which reduces visibility IMHO.

**Comment by CppCXY** [comment]
> > formatting will check to see if it can be converted to double quotes, but diagnostics are simpler.
> > you mentioned `Stylua`, but in the handling of formatting quotes, we are the same.
> > if you think the diagnostic and formatting goals should be consistent at this point, I'm okay with that
> 
> Ty @CppCXY, I’m mentioning stylua just as a reference point, doesn’t mean the behavior should be the same.
> 
> My issue with the diagnostics warning here is that it can never be fixed, if you need the double quotes inside the string you will forever have the warning unless you’re willing to use ugly double quote escape `”b\”a\”r”` which reduces readability IMHO.

you need is that diagnostics should be consistent with the formatting at the quotes.

**Comment by ibhagwan** [comment]
> It's nice that the formatting and diagnostic goals are exactly the same, but it creates a lot of noise.
> currently I also only implement indentation(current version of sumneko has not been updated), whitespace, and quotes on format checking, I think that's enough.

That does simplify things, but as you wrote it also creates noise, if I have warnings I can’t get rid of (without going thorough hoops) than at one point I start to ignore these warnings or worse turn style warnings off.

> you need is that diagnostics should be consistent with the formatting at the quotes.

I’m not sure if I need this warning, if I chose to use single quotes for readability and it cannot easily be converted to double quotes I would assume this is intentional and not expect to see a warning.

In other words, if the formatter decided it’s impossible to automatically convert why warn about it? That could also be considered “consistency” no?

**Comment by CppCXY** [comment]
> In other words, if the formatter decided it’s impossible to automatically convert why warn about it? That could also be considered “consistency” no?

warning it because format diagnostics are a different implementation, and I didn't do the same check at that time

**Comment by CppCXY** [comment]
feature complete

**Comment by ibhagwan** [comment]
> feature complete

Tysm @CppCXY!!
